### PR TITLE
feat: Support cosign --k8s-keychain flag

### LINKED ISSUE #390: connaisseur+cosign on eks: add support for aws credentials
k8s version: 1.20
flavor: eks

**Goal**: 

Use connaisseur+cosign on eks to sign/validate images located in private ecr registry.

**Problem statement**: 

At the moment connaisseur provides 2 methods to auth against a private registry:
(https://github.com/sse-secure-systems/connaisseur/blob/master/helm/values.yaml#L82-L87)
1. A secret containing the registry's credentials (i.e. config.json)
2. A username/password

Since docker doesn't natively support ecr auth, in order to auth to ecr, one needs to use an aws helper:
(https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html)
```
$ TOKEN=$(aws ecr get-login-password --region region)
```
This creates a temporary token valid for 12 hours, which you can then use to authenticate to ecr:
```
$ echo $TOKEN | docker login --username AWS --password-stdin aws_account_id.dkr.ecr.region.amazonaws.com
```

Creating a secret with the following information would unfortunately expire within 12h.
```
{
  "auths": {
    "aws_account_id.dkr.ecr.region.amazonaws.com": {
      "username": "AWS",
      "password": "<TOKEN>"
    }
  }
}
```

The solution I had to come up with as reported by @diegonicacio in https://github.com/sse-secure-systems/connaisseur/issues/352 was to create a cronjob, whose job is to rotate the ecr credentials inside the secret every x hours.

While this works, it's a fragile solution, and it would be a more elegant solution if connaisseur was able to natively support AWS credentials:

```
[default]
aws_access_key_id=AKIAIOSFODNN7EXAMPLE
aws_secret_access_key=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
```

### PR DESCRIPTION
**Pull request by marckn0x**
Fixes #390

## Description

Add configuration option `k8s_keychain` for `cosign` validator that will pass the flag `--k8s-keychain` to `cosign` when set to true. This allows `cosign` to pick up registry credentials from the environment (see https://github.com/sigstore/cosign/pull/972 for details)

## Checklist

- [x] PR is rebased to/aimed at branch `develop`
- [x] PR follows [Contributing Guide](../docs/CONTRIBUTING.md)
- [x] Added tests (if necessary)
- [x] Extended README/Documentation (if necessary)

### CODE REVIEW

**Comment by marckn0x** [comment]
I can add a test if necessary but it's a small change and any test would probably be contrived. Let me know if you think I should.

This is ready for review.

**Comment by marckn0x** [comment]
@phbelitz Doesn't look like I have permission to request a review. Would you mind taking a look at some point?

File: docs/validators/sigstore_cosign.md [hunk 1]
```diff
@@ -97,6 +97,7 @@ kubectl run altsigned --image=docker.io/securesystemsengineering/testimage:co-si
 | `auth.` | | | Authentication credentials for private registries. |
 | `auth.secret_name` | | | Name of a Kubernetes secret in Connaisseur namespace that contains [dockerconfigjson](https://kubernetes.io/docs/concepts/configuration/secret/#docker-config-secrets) for registry authentication. See additional notes [below](#authentication). |
 | `cert` | | | A certificate in PEM format for private registries. |
+| `k8s_keychain` | false | | When true, pass `--k8s-keychain` argument to `cosign verify`. |
```
Comment (xopham) [inline docs/validators/sigstore_cosign.md:100]: key value might need to be adjusted based on the discussion above to `auth.k8s_keychain`. Furthermore, would like to be a bit more explicit in what it does:
```suggestion
| `auth.` | | | Authentication credentials for private registries. See additional notes [below](#authentication). |
| `auth.secret_name` | | | Name of a Kubernetes secret in Connaisseur namespace that contains [dockerconfigjson](https://kubernetes.io/docs/concepts/configuration/secret/#docker-config-secrets) for registry authentication. See additional notes [below](#dockerconfigjson). |
| `auth.k8s_keychain` | false | | When true, pass `--k8s-keychain` argument to `cosign verify` in order to use workload identities for authentication. See additional notes [below](#k8s_keychain). |
| `cert` | | | A certificate in PEM format for private registries. |
```

File: docs/validators/sigstore_cosign.md [hunk 2]
```diff
@@ -136,6 +140,30 @@ kubectl create secret generic my-secret \
 In the above case, the secret name in Connaisseur configuration would be `secret_name: my-secret`.
 It is possible to provide one Kubernetes secret with a `config.json` for authentication to multiple private registries and referencing this in multiple validators.
 
+#### k8s_keychain
+
+Specify `k8s_keychain: true` in the validator configuration to pass the flag `--k8s-keychain` to `cosign` when performing image validation.
+This will allow `cosign` to pick up ambient registry credentials from the environment.
```
Comment (xopham) [inline docs/validators/sigstore_cosign.md:146]: minor proposal:
```suggestion
Specification of `auth.k8s_keychain: true` in the validator configuration passes the `--k8s-keychain` to `cosign` when perfoming image validation.
Thus, [k8schain](https://pkg.go.dev/github.com/google/go-containerregistry/pkg/authn/k8schain) is used by `cosign` to pick up ambient registry credentials from the environment and for example use workload identities in case of common cloud providers.
```
I was unable to find any better documentation of what it is and how it works. Would you be aware of anything?

**Review by xopham — COMMENTED** [review]
@marckn0x this is an awesome feature :tada: Must have slipped my attention, but great that you picked it up :rocket: 

I have some comments. Did you make a functional test of the feature in your environment?
I'd prefer if a small test was added to the pytests, maybe around [here](https://github.com/sse-secure-systems/connaisseur/blob/master/tests/validators/cosign/test_cosign_validator.py#L235), though I admit this maybe of limited use.

**Comment by marckn0x** [comment]
I added a test in the location suggested; thanks!

File: connaisseur/res/config_schema.json [hunk 3]
```diff
@@ -165,6 +165,9 @@
                                 "host": {
                                     "type": "string",
                                     "pattern": "(([-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6})|(localhost))(\\:\\d{2,5})?"
+                                },
+                                "k8s_keychain": {
+                                    "type": "boolean"
```
[the remark this answers is not in the document]
Comment (marckn0x) [reply connaisseur/res/config_schema.json]: Done. I removed the `"required": [ "secret_name" ]` to accommodate passing only `k8s_keychain`.

File: docs/validators/sigstore_cosign.md (hunk 1 above)
Comment (marckn0x) [reply docs/validators/sigstore_cosign.md]: I used your suggested changes. Thanks!

File: docs/validators/sigstore_cosign.md (hunk 2 above)
Comment (marckn0x) [reply docs/validators/sigstore_cosign.md]: I used your suggested changes here.

Unfortunately, the flag doesn't seem to be documented outside the help message displayed by `cosign`, which says (from https://github.com/sigstore/cosign/blob/3ccf3c6a869eaed6329c4e717bfbc7738f0610a1/cmd/cosign/cli/options/registry.go#L49 ): 

`whether to use the kubernetes keychain instead of the default keychain (supports workload identity).`

I think linking to the k8schain documentation, along with the link I added to the cosign pull request implementing the flag, is sufficient for now.

Comment (xopham) [reply docs/validators/sigstore_cosign.md]: agreed

File: connaisseur/res/config_schema.json [hunk 4]
```diff
@@ -152,11 +152,11 @@
                                         "secret_name": {
                                             "type": "string",
                                             "pattern": "[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
+                                        },
+                                        "k8s_keychain": {
+                                            "type": "boolean"
```
Comment (xopham) [inline connaisseur/res/config_schema.json:157]: If I am not mistaken, `k8s_keychain` won't work with `secret_name` as providing both instruction in conjunction would be undefined behavior. Thus, the object should be `OneOf` as used above for the `notary.auth`: https://github.com/sse-secure-systems/connaisseur/pull/551/files#diff-71bf3c9c9646a106b18d918f9cbdd6899845b0d55a2c254f8a7d7ee83149375fL74

**Review by xopham — CHANGES_REQUESTED** [review]
Requested a few last changes and after that we're good to go :rocket:

File: connaisseur/res/config_schema.json (hunk 4 above)
Comment (xopham) [reply connaisseur/res/config_schema.json:157]: would also maintain the `required` if auth is given :slightly_smiling_face:

File: connaisseur/validators/cosign/cosign_validator.py [hunk 5]
```diff
@@ -22,9 +22,10 @@ class CosignValidator(ValidatorInterface):
     name: str
     trust_roots: list
 
-    def __init__(self, name: str, trust_roots: list, **kwargs):
+    def __init__(self, name: str, trust_roots: list, auth: dict = {}, **kwargs):
         super().__init__(name, **kwargs)
         self.trust_roots = trust_roots
+        self.k8s_keychain = auth.get("k8s_keychain", False)
```
Comment (xopham) [inline connaisseur/validators/cosign/cosign_validator.py:28]: function defaults should not be [empty mutables in Python](https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python).
```suggestion
    def __init__(self, name: str, trust_roots: list, auth: dict = None, **kwargs):
        super().__init__(name, **kwargs)
        self.trust_roots = trust_roots
        self.k8s_keychain = False if auth is None else auth.get("k8s_keychain", False)
```

**Review by xopham — CHANGES_REQUESTED** [review]

File: connaisseur/validators/cosign/cosign_validator.py (hunk 5 above)
Comment (xopham) [reply connaisseur/validators/cosign/cosign_validator.py:28]: should be tested when instantiating the class: https://github.com/sse-secure-systems/connaisseur/pull/551/files#diff-60bd9ac24e4f4d47b21f0c14d4ae389b3a7ca9f5b41519d955a12d7c8dcb9d0bR91

I'd propose to test for the auth dict to be:
* `{"k8s_keychain": True}`
* `{"k8s_keychain": False}`
* `{"secret_name": "thesecret"}`

File: connaisseur/res/config_schema.json (hunk 4 above)
Comment (marckn0x) [reply connaisseur/res/config_schema.json]: Done

File: connaisseur/validators/cosign/cosign_validator.py (hunk 5 above)
Comment (marckn0x) [reply connaisseur/validators/cosign/cosign_validator.py]: Fixed the default, thanks for that!

I modified the test to pass `auth` as each of these three cases. I'm not sure what you mean by "should be tested when instantiating the class: <link>", as <link> doesn't jump to any specific diff on this PR for me.

Anyway, the entire statement is now covered by the test.

Comment (marckn0x) [reply connaisseur/validators/cosign/cosign_validator.py]: Added one more test case for `auth = None`. Now the entire statement is really covered :)

Comment (xopham) [reply connaisseur/validators/cosign/cosign_validator.py]: I made a fixup commit for the class instantiation test. Think my wording was unclear. Would that work for you?

**Review by xopham — COMMENTED** [review]
I made a proposal for the tests. If that works for you, feel free to squash and I'd be happy to approve :slightly_smiling_face:

File: connaisseur/validators/cosign/cosign_validator.py (hunk 5 above)
Comment (marckn0x) [reply connaisseur/validators/cosign/cosign_validator.py]: Works for me. I just squashed your fixup commit, thanks!

**Review by xopham — APPROVED** [review]

**Review by xopham — APPROVED** [review]
@marckn0x thanks for the great contribution :rocket:
